Fix: load scope as xform (FTV-417)

judubu opened the pull request:
Fix https://github.com/Unity-Technologies/usd-unity-sdk/issues/154

Scopes are not Xformable per se but can contain payload and hold variantSets. If we want to be able to set the variant before loading the payload we need to load the Scope as a xform.

judubu committed:
Fix: load scope as xform (FTV-417)

https://github.com/Unity-Technologies/usd-unity-sdk/issues/154

Scopes are not Xformable per se but can contain payload and hold variantSets. If we want to be able to set the variant before loading the payload we need to load the Scope as a xform.

thomas-tu reviewed (commented):
Should the `ScopeSample` array be injected in the PrimMap like the others results?

thomas-tu commented:
Additionally, could we maybe add tests to make sure we load things properly with different cases?

judubu commented:
@thomas-tu The only reason we'd want the Scope prims added to the primMap would be to run specific import/export logic on them. Given that Scopes are represented in Unity as Xforms and that  the exporter automatically exports the whole parent hierarchy for any valid prims I don't think we need to add specific ScopeImporter/Exporter for now. https://github.com/Unity-Technologies/usd-unity-sdk/blob/b28f0fee38a2a286e199d1f6e0d260789f475279/package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/SceneExporter.cs#L635-L637
That will be required if/when we focus on proper round tripping of USD assets.

judubu committed:
Add comment explaining why Scopes are not added to the PrimMap

judubu committed:
Add tests

judubu committed:
Update Usd.Tests asmdef

judubu committed:
Fix test data path

jcowles commented on package/com.unity.formats.usd/Tests/Editor/EditorTests.cs:
```diff
@@ -1,4 +1,7 @@
 using NUnit.Framework;
+using UnityEngine;
+using USD.NET;
+using System.IO;
```
AMAZING, A TEST!

jcowles commented on package/com.unity.formats.usd/Tests/Editor/EditorTests.cs:
```diff
@@ -9,5 +12,35 @@ public void DummyTest()
+            InitUsd.Initialize();
+            var stage = pxr.UsdStage.Open(Path.Combine(Application.dataPath, filepath), pxr.UsdStage.InitialLoadSet.LoadNone);
+            var scene = Scene.Open(stage);
+            usdRoot = USD.UsdMenu.ImportSceneAsGameObject(scene);
```
Should probably make a SceneLoader utility class (either in this change or a future change when adding new tests), so all tests don't have to implement this logic

judubu replied:
That's what I've been telling myself for each branch I've worked on so far... 
I need to rewrite the loading part anyway because assets are not found on Yamato. Adding them to the package and accessing them through guid seems to be the correct way.

thomas-tu replied:
+1. 

You can make a base framework to do that for you. That will probably become very redundant. Plus, you delegate the knowledge of "How do I load my stuff on Yamato" to one place, allowing people to focus on their tests rather than Yamato-specific rules.

judubu replied:
I'll create a separate task for that

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Scene/HierarchyBuilder.cs:
```diff
@@ -151,6 +151,7 @@ public struct Query<T> : IQuery where T : SampleBase, new() {
       }
       if (options.importTransforms) {
         FindPathsJob.queries[7] = (FindPathsJob.IQuery)new FindPathsJob.Query<XformSample>();
+        FindPathsJob.queries[8] = (FindPathsJob.IQuery)new FindPathsJob.Query<ScopeSample>();
```
We may need to import scopes regardless of transform import state... I don't know if there is a use case, but I could imagine making a hierarchy of scopes, disabling `importTransforms` and then being confused why my scopes are showing up.

jcowles commented on package/com.unity.formats.usd/Tests/Editor/EditorTests.cs:
```diff
@@ -9,5 +12,35 @@ public void DummyTest()
+        public void TestScopeWithoutChildren()
+        {
+            var scopeWithChildren = usdRoot.transform.Find("TestComponent/geom");
+            Assert.IsNotNull(scopeWithChildren);
```
Do we want to assert anything about the data on the scope? E.g. that it has a UsdPrim component or that the transform is identity?

thomas-tu replied:
@jcowles, is it valuable for this test? Or should they be in separate ones?

judubu replied:
You're right, I'll do both and add a little bit more because I'm realizing the test doesn't do much. Looks like I lost some bits between stashes...

jcowles reviewed (approved):
Nice, love that you added a real test!

Some minor comments, but mostly just for your consideration.

thomas-tu commented on package/com.unity.formats.usd/Tests/Editor/EditorTests.cs:
```diff
@@ -9,5 +12,35 @@ public void DummyTest()
+        }
+        
+        [Test]
+        public void TestScopeWithoutChildren()
```
We should apply some rules on the test naming. You would like to be able what broke from a log without to have to open the code. My rule of thumb: `<LogicName/MethodName>_<Input>_<ExpectedOutput>`.

Names become long but this is necessary to prevent time consuming back and forth between Yamato and a local repository.

thomas-tu commented on package/com.unity.formats.usd/Tests/Editor/EditorTests.cs:
```diff
@@ -9,5 +12,35 @@ public void DummyTest()
+        }
+        
+        [Test]
+        public void TestScopeWithChildren()
```
See my comment about `TestScopeWithoutChildren()`:


> We should apply some rules on the test naming. You would like to be able what broke from a log without to have to open the code. My rule of thumb: <LogicName/MethodName>_<Input>_<ExpectedOutput>.
> 
> Names become long but this is necessary to prevent time consuming back and forth between Yamato and a local repository.

thomas-tu reviewed (changes requested):
Nice! Looks good!

I left a comment about naming tests. I think that will be important in the future.
Also, do you expect to re-use the `.usda` files in differents tests? I usually classify my assets in folder that is also name after the tests main class for tracking purposes. In this case, they would have been put in `ScopeLoadingTests/`.

judubu committed:
Add more tests

judubu committed:
Fix tests assertions

judubu committed:
Revert "Update Usd.Tests asmdef"

This reverts commit 4053d6c1

judubu committed:
CI: use package-ci/mac image instead of buildfarm/mac

Fix the following errors:
```
[12:31:47.729 INF] UPM-CI is not compatible with your installed version of Nodejs (v9.8.0). Please use Nodejs v10 or above.
For more information on how to resolve this error, including in CI: https://github.cds.internal.unity3d.com/unity/upm-ci-utils#upm-ci-is-not-compatible
---------
/usr/local/lib/node_modules/upm-ci-utils/scripts/postinstall.js:9
    throw new Error(`Incompatible Nodejs Version (${process.version})`);
    ^

Error: Incompatible Nodejs Version (v9.8.0)
    at checkEngineVersion (/usr/local/lib/node_modules/upm-ci-utils/scripts/postinstall.js:9:11)
    at Object.<anonymous> (/usr/local/lib/node_modules/upm-ci-utils/scripts/postinstall.js:13:1)
    at Module._compile (module.js:649:30)
    at Object.Module._extensions..js (module.js:660:10)
    at Module.load (module.js:561:32)
    at tryModuleLoad (module.js:501:12)
    at Function.Module._load (module.js:493:3)
    at Function.Module.runMain (module.js:690:10)
    at startup (bootstrap_node.js:194:16)
    at bootstrap_node.js:666:3
```

judubu committed:
CI: updating promotion.yml with package-ci.mac image

judubu commented:
The yamato failure on trunk comes from the dev branch. Investigation in progress. Meanwhile I'm merging this branch which has been approved for ages and had its trunk tests passing before.
